Preliminary "foot in the door" SQLite support

brandur opened the pull request:
Here, add an implementation for SQLite. The new driver passes the entire
driver test suite (including some newly added tests), and checks out
against a new end-to-end suite of client tests targeting it. I think at
this point we can consider it to be largely functional, although with
minimal real-world vetting, but at least alpha quality that we can start
putting in front of users.

The main reason this turned out to be quite an effort is that SQLite,
although nominally supporting a lot of the same syntax Postgres does,
just has a lot of unexpected limitations and quirks that all need to be
tracked down separately and a workaround added. For example:

* SQLite doesn't really have data types. Everything is either an
  integer, real, text, or blob, and anything more sophisticated just has
  to be shoehorned into one of these types. For example, there's no
  boolean. Despite the fanfare recently, there isn't even a jsonb. Jsonb
  gets pushed into a blob.

* The most annoying missing data type is a timestamp. Date/times are
  stored as either Unix integers or strings, and SQLite provides a
  number of obnoxious, error-prone, built-ins to work with them.

* No arrays.

* No modification operations in CTEs. (Although counting my lucky stars
  that CTEs are supported at least.)

* No listen/notify.

Aside from that though, we had to get some testing infrastructure in
place, and our testing has traditionally been quite affixed to Postgres,
which would've made a lot of changes necessary. However, the schema
additions in #848 do help to make this quite a bit easier with the
driver-based `TestSchema` and `TestTx` functions, and because they raise
schema automatically, it doesn't complicate test setup instructions in
any way by adding extra steps.

A new `TestDriverRiverSQLite` test has been added to run the full
barrage of driver tests for SQLite. I've also added a new family of
`TestClientWithDriver` tests that are similar to `TestDriver*` in that
they run for each of our supported drivers, but in this case they do a
series of end-to-end tests to more fully exercise the entire River
client. I added a respectable set of base client driver tests that walks
each driver through all basic functionality (including for features that
are higher risk of failing from one database compared to the next like
`JobList`), but as with most tests, they could stand to be extended
further, which I'm sure we'll get to in the future.

An example test for SQLite has been added, demonstrating the use of
River with a SQLite driver and in-memory SQLite database.

A few caveats:

* I've found that the combination of SQLite's reduced capabilities +
  sqlc's bugginess [1] make bit batch insert/update operations basically
  impossible (I tried every workaround I could possibly thin of), so
  these are done in loops of individual operations instead. I *think*
  this is okay for now though. For one, sqlc will hopefully get these
  limitations fixed eventually, and for two, SQLite databases will often
  be running locally, meaning the round trip cost per operation is much
  lower than what we'd see in a hosted Postgres somewhere.

* It's not totally clear that having a SQLite implementation will be
  "worth it enough" in the long run given that it will add some
  difficulty to non-trivial future database-related operations. My hope
  is that as it's in a prerelease state, we can gauge how bad it is to
  keep up to date. If it feels like way more effort than it's worth, we
  can still axe it before it ever becomes a functional driver.

[1] https://github.com/sqlc-dev/sqlc/issues/3802#issuecomment-2835863277

brandur commented:
@travisby Great! Thanks for the feedback. We'll try to get this over the line.

brandur commented:
Alright, I added implementations for all of `river_queue` as well, along with `river_client` + `river_client_queue` although these aren't actually used and we'll probably delete them. That leaves only `river_job` left.

I ran into a bit of a problem in that I'm pretty sure sqlc + SQLite + use of jsonb is completely broken, and had to implement everything using normal json for now (which TBH, probably isn't that much worse). I opened a discussion about that on sqlc, but given some of the track record of other questions/comments I've left on sqlc issues over the last couple weeks, I'm not super optimistic that it'll go anywhere:

https://github.com/sqlc-dev/sqlc/discussions/3953

I did fix at least one existing bug during the implementation too, namely that `river_queue` being unpaused would target an already unpaused queue despite comments in tests clearly indicating that this was unintentional (and missing assertions to verify a no-op), which is kind of a nice side benefit.

bgentry commented on riverdriver/riversqlite/go.mod:
```diff
@@ -0,0 +1,18 @@
+module github.com/riverqueue/river/riverdriver/riversqlite
+
+go 1.23.0
+
+toolchain go1.24.1
+
+require (
+	github.com/riverqueue/river/riverdriver v0.20.2
+	github.com/riverqueue/river/rivershared v0.20.2
+	github.com/riverqueue/river/rivertype v0.20.2
+	github.com/stretchr/testify v1.10.0
+)
+
+require (
+	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/pmezard/go-difflib v1.0.0 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
+)
```
Shouldn't this driver module actually require the sqlite library i.e. `modernc.org/sqlite`?

brandur replied:
So believe it or not ... no. All code in the driver itself only depends on `sqlc` primitives. That's actually the case for `riverdriversql` as well, except that sqlc does need to pull in pq for a couple small things like the `pq.Array` helper.

We need an underlying driver, but only where we're actually running tests, and just the way the project is laid out, it turns out that all the tests are in the top-level package to prevent package dependency cycle problems (e.g. see `driver_test.go`). It would actually be nice to see if we could take another stab at pushing tests into each specific driver maybe, and if we pulled that then this package would pick up the `modernc.org/sqlite` dependency.

brandur replied:
Also, I ran `go mod tidy` on all these files now, so the dependencies should be a little more correct now.

bgentry commented on riverdriver/riversqlite/internal/dbsqlc/river_job.sql:
```diff
@@ -0,0 +1,426 @@
+-- Instead, the driver uses a transaction to optimisticaly try an update, but
+-- perform a subsequent fetch on a not found to return the right status.
+--
+-- I had to invert the last 'AND' expression below (was an 'ANT NOT) due to an
```
```suggestion
-- I had to invert the last 'AND' expression below (was an 'AND NOT') due to an
```

brandur replied:
Thx. Fixed.

bgentry commented on riverdriver/riversqlite/internal/dbsqlc/river_job.sql:
```diff
@@ -0,0 +1,426 @@
+ORDER BY /* TEMPLATE_BEGIN: order_by_clause */ id /* TEMPLATE_END */
+LIMIT @max;
+
+-- Rescue a job.
+--
+-- This is supposed to be a batch insert, but various limitations of the
+-- combined SQLite + sqlc has left me unable to find a way of injecting many
+-- arguments en masse (like how we slightly abuse arrays to pull it off for the
+-- Postgres drivers), so we loop over many insert operations instead, with the
+-- expectation that this may be fixable in the future. Because SQLite targets
+-- will often be local and therefore with a very minimal round trip compared to
+-- a network, looping over operations is probably okay performance-wise.
```
Comment not applicable to this query?

brandur replied:
I had copied it from another query, but it still is somewhat relevant in notes on what makes it different from the Postgres `JobRescue` implementation. I just tailored it a bit better to this specific query.

bgentry commented on riverdriver/riversqlite/internal/dbsqlc/river_leader.sql:
```diff
@@ -0,0 +1,76 @@
+FROM /* TEMPLATE: schema */river_leader
+WHERE leader_id = @leader_id;
+
+-- WITH currently_held_leaders AS (
+--   SELECT *
+--   FROM /* TEMPLATE: schema */river_leader
+--   WHERE leader_id = @leader_id::text
+--   FOR UPDATE
+-- ),
+-- notified_resignations AS (
+--     SELECT pg_notify(
+--         concat(coalesce(sqlc.narg('schema')::text, current_schema()), '.', @leadership_topic::text),
+--         json_build_object('leader_id', leader_id, 'action', 'resigned')::text
+--     )
+--     FROM currently_held_leaders
+-- )
+-- DELETE FROM /* TEMPLATE: schema */river_leader USING notified_resignations;
```
want to leave a comment on this?

brandur replied:
Ah actually, that was just scratch I was using as a reference while writing the new query. I just removed it. Thx.

bgentry commented on riverdriver/riversqlite/river_sqlite_driver.go:
```diff
@@ -0,0 +1,1415 @@
+
+			tags, err := json.Marshal(params.Tags)
+			if err != nil {
+				// return nil, err
```
```suggestion
```

brandur replied:
Thx. Removed.

bgentry reviewed (approved):
Looks solid! My comments are primarily about non-sqlite things that are being made worse by this and wanting to avoid those if possible, but otherwise we should move forward with this and see how it goes.

brandur commented:
Thx Blake.

I did some more work on this one and I think I have it at pretty much a fully functional driver. Added a bunch more driver client tests and an example test demonstrating its use. I think given the good test coverage, we could at least start putting this in front of people in an alpha form.

Also found a bug in `TestSchema` where schemas weren't being checked back in for reuse after being reused once, so the net effect is that this PR makes the main package's test suite another 1-2s faster locally. (Down from ~10.5s to ~8.5s, but timing varies a lot so +/-.)

I addressed some of your feedback but didn't respond yet. I'll crunch through it all in the next day or two.

brandur commented on client.go:
```diff
@@ -246,6 +246,11 @@ type Config struct {
+	// QueuePollInterval is the amount of time between periodic checks for queue
+	// setting changes. This is only used in poll-only mode (when no notifier is
+	// provided).
+	QueuePollInterval time.Duration
```
Okay, lemme know if you disagree, but although I left this new addition in `Config`, I flipped it to being unexported since we haven't seen any major demand for it yet.

I had to add it for the good of the test suite so that queue-based tests with a poll-only driver weren't too slow, but probablyyy most people don't need this. Better to keep the API smaller for the time being unless someone asks for it.

brandur commented on client.go:
```diff
@@ -1865,20 +1871,23 @@ func (c *Client[TTx]) maybeNotifyInsertForQueues(ctx context.Context, tx riverdr
-			slog.String("err", err.Error()),
-		)
-		return err
+	if c.driver.SupportsListenNotify() {
```
100%.

I thought about this a lot yesterday, maybe less so for what you're talking about, but mainly more in pursuit of finding ways to speed up tests where poll-only is active. The transactions make things difficult, but I thinkkk there's something we could do in adding something like an emit hook post-transaction-commit on the non-transaction function variants like `Insert` or `QueuePause`.

I almost took a stab at implementing this, but realized it's way too big to try and shoehorn in with the rest of these changes, so I stopped.

I'm interested in working on this soon though — like maybe next.

brandur commented on driver_client_test.go:
Awesome, thx. Yeah it should reveal quite a few problems that don't currently get exposed.

brandur commented on internal/dblist/db_list.go:
```diff
@@ -60,38 +61,56 @@ func JobList(ctx context.Context, exec riverdriver.Executor, params *JobListPara
+				whereBuilder.WriteString(",")
+			}
+
+			argName := fmt.Sprintf("list_arg_%02d", argNum)
```
Yeah good point.

Okay now that I have more testing in place, I went back to the drawing board and rebuilt this completely by pushing specific implementations for these clauses back down into drivers.

Postgres implementation basically goes back to the way it was:

``` go
func (d *Driver) SQLFragmentColumnIn(column string, values any) (string, any, error) {
	return fmt.Sprintf("%s = any(@%s)", column, column), values, nil
}
```

And I was able to tighten up the SQLite implementation to be pretty close by using `json_each`:

``` go
func (d *Driver) SQLFragmentColumnIn(column string, values any) (string, any, error) {
	arg, err := json.Marshal(values)
	if err != nil {
		return "", nil, err
	}

	return fmt.Sprintf("%s IN (SELECT value FROM json_each(cast(@%s AS blob)))", column, column), arg, nil
}
```

(`json_each` is a really handy primitive for SQLite and should be able to support a lot of our batch operations, but unfortunately can't right now because sqlc's support for it is absolute garbage. I was able to use it here though since we're just building a string fragment anyway, so sqlc is out of the loop. I'd like to go and see if I can fix some sqlc parsing upstream and then maybe go back in and use it in more of the SQLite driver for River.)

Each invocation in `dblist` becomes something like this:

``` go
	if len(params.IDs) > 0 {
		writeAndAfterFirst()

		const column = "id"
		sqlFragment, arg, err := sqlFragmentColumnIn(column, params.IDs)
		if err != nil {
			return nil, fmt.Errorf("error building SQL fragment for %q: %w", column, err)
		}
		whereBuilder.WriteString(sqlFragment)
		namedArgs[column] = arg
	}
```

So the N parameters go away in favor of something that's quite close to what we had before.

brandur commented on internal/riverinternaltest/riverdrivertest/riverdrivertest.go:
```diff
@@ -1025,13 +1257,153 @@ func Exercise[TTx any](ctx context.Context, t *testing.T,
+				require.Equal(t, 1, rowsAffected2)
+			default:
+				require.FailNow(t, "Unknown database name: "+bundle.driver.DatabaseName())
+			}
```
Yeah good call.

Actually, going back in to look at this again, it turned out to be a mistake on my part. In SQLite I'd copy/pasta'ed the two `JobInsertMany` + `JobInsertManyNoReturning` queries, but forgotten to add `DO NOTHING` onto the second one, which is why the results were coming back differently.

I fixed up that mistake and was able to remove all this test code.

brandur commented on internal/riverinternaltest/riverdrivertest/riverdrivertest.go:
```diff
@@ -2187,78 +2639,103 @@ func Exercise[TTx any](ctx context.Context, t *testing.T,
 
 	const leaderTTL = 10 * time.Second
 
+	const veryGenerousTimeCompareTolerance = 55 * time.Minute
```
Haha. Actually I can't remember why I set it _that_ high. Changed to 5 minutes, which should be plenty.

brandur commented on internal/riverinternaltest/riverdrivertest/riverdrivertest.go:
```diff
@@ -1886,6 +2325,19 @@ func Exercise[TTx any](ctx context.Context, t *testing.T,
+			// of an unknown job. This may not have been a particularly well
+			// thought out decision, but I've codified here later to help make
+			// sure all drivers behave consistently.
+			_, err := exec.JobSetStateIfRunningMany(ctx, setStateManyParams(riverdriver.JobSetStateCompleted(0, time.Now().UTC(), nil)))
```
Okay good point. I updated the comment to better reflect that:

``` go
			// The operation doesn't return anything like a "not found" in case
			// of an unknown job so that it doesn't fail in case a job is
			// deleted in the interim as a completer is trying to finalize it.
```

brandur commented on riverdriver/riverdatabasesql/river_database_sql_driver.go:
```diff
@@ -708,6 +629,10 @@ func (e *Executor) MigrationGetByLine(ctx context.Context, params *riverdriver.M
 func (e *Executor) MigrationInsertMany(ctx context.Context, params *riverdriver.MigrationInsertManyParams) ([]*riverdriver.Migration, error) {
+	// ctx = sqlctemplate.WithReplacements(ctx, map[string]sqlctemplate.Replacement{
+	// 	"version": {Value: strings.Join(sliceutil.Map(params.Versions, func(v int) string { return strconv.Itoa(v) }), ", ")},
+	// }, nil)
```
Good catch. Removed.

brandur commented on riverdriver/riverpgxv5/internal/dbsqlc/river_queue.sql:
```diff
@@ -21,15 +21,15 @@ INSERT INTO /* TEMPLATE: schema */river_queue(
 ) ON CONFLICT (name) DO UPDATE
 SET
-    updated_at = coalesce(sqlc.narg('updated_at')::timestamptz, sqlc.narg('now')::timestamptz, now())
+    updated_at = EXCLUDED.updated_at
```
Sweet. Yep, I think this was just an accidental omission from before.

brandur commented on riverdriver/riverpgxv5/internal/dbsqlc/river_job.sql:
```diff
@@ -379,9 +371,7 @@ WITH jobs_to_schedule AS (
     WHERE
         state IN ('retryable', 'scheduled')
-        AND queue IS NOT NULL
-        AND priority >= 0
```
Hmm, okay I'm fairly sure that it does nothing because you're asserting on conditions that are always true so this is functionally an `AND true AND true`, but admittedly this tweak has nothing to do with SQLite, so I just reverted it. We can possibly test it again elsewhere.

brandur commented:
@bgentry Alright, I think all code review feedback has been addressed and we're at a point where this change should be quite functional. Test suite stability is extremely good. Want to do one more pass?

brandur commented on client.go:
```diff
@@ -715,7 +725,7 @@ func NewClient[TTx any](driver riverdriver.Driver[TTx], config *Config) (*Client
 	// them):
 	if config.willExecuteJobs() {
-		if !driver.HasPool() {
+		if !driver.PoolIsSet() {
```
I ended up renaming this so that it sorts more closely to its with `PoolSet` in the driver files.

bgentry commented on internal/riverinternaltest/riverdrivertest/riverdrivertest.go:
```diff
@@ -216,41 +414,49 @@ func Exercise[TTx any](ctx context.Context, t *testing.T,
+		// it back early means that a detach can no longer be run on it. It
+		// doesn't matter much if an extra schema is attached until the tests
+		// finish running.
+		// require.NoError(t, exec.SchemaAttach(ctx, &riverdriver.SchemaAttachParams{Schema: schemaVersion2}))
```
making sure you meant to leave this in here

brandur replied:
Ah nope. That was part of a previous attempt in getting things to work and no longer needed. I also removed one other similar block in this file that was also commented out.

bgentry reviewed (approved):
made it through again 😅

brandur commented on rivershared/go.mod:
```diff
@@ -12,16 +12,27 @@ require (
 	github.com/stretchr/testify v1.10.0
 	go.uber.org/goleak v1.3.0
 	golang.org/x/mod v0.24.0
+	modernc.org/sqlite v1.37.0
```
Doh, damn, good catch on that. Yeah, that would've added modernc is a dep everywhere.

Basically what happened is that as I was trying to test things, I'd added an `import ( _ "modernc.org/sqlite" )` to `riversharedtest`. It wasn't needed by the end, but because of the underscore import, goimport can't detect that it'd not being used, and therefore left it in.

There is a `DBPoolSQLite` function in there, but it only needs to use the `sql` package. The caller tests must be importing `modernc.org/sqlite`, but that's in the test assembly.

I took out the import and tidied. This dependency is gone now.

brandur commented on client.go:
```diff
@@ -1997,12 +2014,16 @@ func (c *Client[TTx]) JobList(ctx context.Context, params *JobListParams) (*JobL
 	params.schema = c.config.Schema
 
+	if c.driver.DatabaseName() == "sqlite" && params.metadataFragment != "" {
+		return nil, errors.New("JobListResult.Metadata is not supported on SQLite")
```
Let's discuss this more, but I'm kind of thinking we should switch to JSON Path for that stuff. It's more general and can be supported by multiple databases.

brandur commented on go.mod:
```diff
@@ -21,15 +23,30 @@ require (
+	modernc.org/memory v1.10.0 // indirect
 )
+
+// TODO(brandur): Remove this before first release including SQLite driver. It's
+// needed temporarily because there's no riversqlite tag to target (the one
+// referenced above is fake and does not exist).
+replace github.com/riverqueue/river/riverdriver/riversqlite => ../river/riverdriver/riversqlite
```
Yeah, +1. I have a big warning banner at the top of my TODO list as a reminder to myself to remove it.

brandur commented on CHANGELOG.md:
```diff
@@ -7,6 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+### Added
+
+- Preliminary River driver for SQLite (`riverdriver/riversqlite`). This driver seems to produce good results as judged by the test suite, but so far as minimal real world vetting. Try it and let us know how it works out. [PR #870](https://github.com/riverqueue/river/pull/870).
```
Thx. Fixed.

brandur commented:
Thanks Blake!

Okay, I changed my mind a bit and am going to merge this in after all. I do still want to try and get an improvement for jsonb in there, but even if I pull that off, the sqlc release cycles are so long that it's going to take some time. I also realized that since json and jsonb are both actually `blob` in SQLite, there won't be any required DDL for anyone who made a table with json. Given the driver is still pretty alpha, I think I'll just add a note in the changelog with some `UPDATE` statements that'll convert json to jsonb instead of doing a full migration.

I also added SQLite support to the River CLI to make sure people try to use it have the ability to run migrations and everything.
